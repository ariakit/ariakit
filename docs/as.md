`as` is one of the most powerful functions of ReaKit. This leverages composability in React and provides a nice way to create new components by composing many others.

Your first contact with `as` will be most probably with the prop, changing the underlying element of a component:
```jsx
import { Button } from "reakit";

<Button as="a" href="https://github.com/reakit/reakit" target="_blank">
  Go to GitHub
</Button>
```

You can pass HTML string elements, other React components or an array of both.

You can use the `as` enhancer to create new components and take advantage of all ReaKit features:

```jsx
import { as, Button } from 'reakit'

const MyComponent = as("span")(
  ({ as: T, ...props }) => <T {...props} />
);

<MyComponent 
  as={[Button, "a"]} 
  href="https://github.com/reakit/reakit" 
  target="_blank"
>
  GitHub
</MyComponent>
```

That means you can also use it to enhance a ReaKit component with `as` and change what it renders:

```jsx
import { as, Button } from 'reakit'

const LinkButton = as("a")(Button);

<LinkButton href="https://github.com/reakit/reakit" target="_blank">
  GitHub
</LinkButton>
```

Finally, all enhanced components expose an `as` static method, which has the same API as the other versions:

```jsx
import { Button } from "reakit";

const LinkButton = Button.as("a");

<LinkButton href="https://github.com/reakit/reakit" target="_blank">
  GitHub
</LinkButton>
```