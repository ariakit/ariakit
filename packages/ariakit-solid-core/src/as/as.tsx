import { combineProps } from "@solid-primitives/props";
import type { Component, ComponentProps, JSX, ValidComponent } from "solid-js";
import { Dynamic } from "solid-js/web";

type AsElements = {
  [K in keyof JSX.IntrinsicElements]: Component<ComponentProps<K>>;
};

type AsComponent = <T extends Component<any>, P = ComponentProps<T>>(
  props: AsProps<T, P>,
) => JSX.Element;

const cache = new Map<string, Component<any>>();

/**
 * Allows a component to be rendered as a different HTML element or Solid
 * component. Must be passed to the `render` prop of a component that
 * supports it.
 *
 * To render as an HTML element, use `<As.element />` (e.g. `<As.button />`).
 *
 * To render as a component, use `<As component={Component} />` (e.g. `<As
 * component={MyButton} />`).
 *
 * Check out the [Composition](https://solid.ariakit.org/guide/composition)
 * guide for more details.
 * @example
 * ```jsx
 * <Role render={<As component={MyButton} variant="primary" />} />
 * <Role render={<As.button type="button" />} />
 * ```
 */
export const As = new Proxy(
  function As(props: any) {
    return ((parentProps: unknown) => (
      <Dynamic
        {...combineProps([parentProps, props], { reverseEventHandlers: true })}
        component={props.component}
      />
    )) as unknown as JSX.Element;
  } as AsComponent & AsElements,
  {
    get: (_, key: keyof JSX.IntrinsicElements) => {
      let component = cache.get(key);
      if (!component) {
        component = function AsElement(props: any): JSX.Element {
          return ((parentProps: unknown) => (
            <Dynamic
              {...combineProps([parentProps, props], {
                reverseEventHandlers: true,
              })}
              component={key}
            />
          )) as unknown as JSX.Element;
        };
        cache.set(key, component);
      }
      return component;
    },
  },
);

export type AsProps<T extends ValidComponent, P = ComponentProps<T>> = {
  [K in keyof P]: P[K];
} & { component: T };
