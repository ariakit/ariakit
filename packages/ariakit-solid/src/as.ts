export { As } from "@ariakit/solid-core/as/as";

export type { AsProps } from "@ariakit/solid-core/as/as";
