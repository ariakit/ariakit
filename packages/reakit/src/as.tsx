/* eslint-disable no-param-reassign */
import * as React from "react";
import hoistNonReactStatics from "hoist-non-react-statics";
import pickCSSProps from "./_utils/pickCSSProps";
import parseTag from "./_utils/parseTag";
import dedupeClassName from "./_utils/dedupeClassName";
import pickHTMLProps from "./_utils/pickHTMLProps";
import getComponentName from "./_utils/getComponentName";
import toArray from "./_utils/toArray";
import omit from "./_utils/omit";
import { AsProps, Dictionary, AsElement, AsComponent } from "./_utils/types";

function As(props: AsProps<any>) {
  return render(Object.assign({}, omit(props, "nextAs"), { as: props.nextAs }));
}

function render(props: AsProps<any>) {
  const T = parseTag(props.as);

  if (Array.isArray(T)) {
    const [First, ...others] = T.filter(x => x !== As);
    const other = others.length === 1 ? others[0] : others;
    // @ts-ignore: We can't be sure if First accepts `as` or `nextAs`
    return <First {...props} as={As} nextAs={other} />;
  }

  const style = pickCSSProps(props);

  if (typeof T === "string") {
    const className = dedupeClassName(props.className);
    const allProps = Object.assign(
      pickHTMLProps(props),
      { className },
      style ? { style } : {}
    );

    return <T {...allProps} ref={props.elementRef} />;
  }

  const allProps = Object.assign(omit(props, "as"), style ? { style } : {});
  return <T {...allProps} />;
}

function isAsComponent(target: any): target is AsComponent<any, any> {
  return typeof target.asComponents !== "undefined";
}

function as<T extends AsElement<any>>(asComponents: T | T[]) {
  return <P extends Dictionary>(WrappedComponent: React.ComponentType<P>) => {
    // Transform WrappedComponent into ReakitComponent
    const defineProperties = (scope: AsComponent<T, P>) => {
      scope.asComponents = asComponents;
      // @ts-ignore
      scope.as = otherComponents => as(otherComponents)(scope);
      return scope;
    };

    // WrappedComponent was already enhanced with the same arguments
    if (
      isAsComponent(WrappedComponent) &&
      asComponents === WrappedComponent.asComponents
    ) {
      return defineProperties(WrappedComponent as AsComponent<T, P>);
    }

    const componentName = getComponentName(WrappedComponent);
    const commaSeparatedAs = ([] as any[])
      .concat(asComponents)
      .map(getComponentName);
    const displayName = `${componentName}.as(${commaSeparatedAs})`;

    const EnhancedComponent = (props =>
      render(
        Object.assign({}, props, {
          as: [
            WrappedComponent,
            ...toArray(asComponents),
            ...toArray(props.as || []),
            ...toArray(props.nextAs || [])
          ]
        })
      )) as AsComponent<T, P>;

    EnhancedComponent.displayName = displayName;
    // @ts-ignore: Only docs
    EnhancedComponent.propTypes = WrappedComponent.propTypes;
    // @ts-ignore: Only docs
    EnhancedComponent.defaultProps = WrappedComponent.defaultProps;
    // @ts-ignore
    hoistNonReactStatics(EnhancedComponent, WrappedComponent);

    return defineProperties(EnhancedComponent);
  };
}

export default as;
