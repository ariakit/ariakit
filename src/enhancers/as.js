import React from 'react'
import styled from 'styled-components'
import omit from 'lodash/omit'
import { pickHTMLProps, pickSVGProps } from 'pick-react-known-prop'
import pickCSSProps from '../utils/pickCSSProps'
import isSVGElement from '../utils/isSVGElement'
import parseTag from '../utils/parseTag'
import parseClassName from '../utils/parseClassName'
import escapeClassName from '../utils/escapeClassName'

const Next = ({ nextAs, ...props }) => <Reas {...props} as={nextAs} />

const Reas = ({ as: t, ...props }) => {
  const T = parseTag(t)

  if (Array.isArray(T)) {
    const FirstT = T[0]
    return <FirstT as={Next} {...props} nextAs={T.slice(1)} />
  }

  const style = pickCSSProps(props)
  const className = parseClassName(props.className)

  if (typeof T === 'string') {
    const { dangerouslySetInnerHTML, children } = props
    const propsWithoutStyle = omit(props, Object.keys(style))
    const propsWithStyle = {
      ...propsWithoutStyle,
      ...(style ? { style } : {}),
    }
    const otherProps = dangerouslySetInnerHTML
      ? { dangerouslySetInnerHTML }
      : {}
    const allProps = {
      ...(isSVGElement(T) && pickSVGProps(propsWithStyle)),
      ...pickHTMLProps(propsWithStyle),
      ...otherProps,
      className,
    }
    return <T {...allProps}>{children}</T>
  }

  return <T {...props} className={className} {...(style ? { style } : {})} />
}

const getComponentName = component =>
  component.displayName || component.name || component

const getDisplayName = ([Component, ...elements]) =>
  `${getComponentName(Component)}.as(${elements.map(getComponentName)})`

const getAs = (components, props) =>
  components.concat(props.as || [], props.nextAs || [])

const as = asComponents => WrappedComponent => {
  const components = [].concat(WrappedComponent, asComponents)

  const AsReas = props => (
    <Reas {...props} as={getAs(components, props)} nextAs={undefined} />
  )

  AsReas.displayName = getDisplayName(components)

  const StyledAsReas = styled(AsReas)`
    margin: 0;
    padding: 0;
    border: 0;
    font-size: 100%;
    font-family: inherit;
    vertical-align: baseline;
    box-sizing: border-box;
  `
  StyledAsReas.displayName = `styled(${AsReas.displayName})`
  StyledAsReas.styledComponentId = escapeClassName(StyledAsReas.displayName)

  StyledAsReas.propTypes = components.reduce(
    (finalPropTypes, component) => ({
      ...finalPropTypes,
      ...component.propTypes,
    }),
    {},
  )

  StyledAsReas.as = otherElements => as(otherElements)(StyledAsReas)

  const originalExtend = StyledAsReas.extend

  Object.defineProperty(StyledAsReas, 'extend', {
    value: (strings, ...interpolations) => {
      const modifiedStrings = [
        `&& {${strings[0] || ''}`,
        ...strings.slice(1, -1),
        `${strings[strings.length - 1] || ''}}`,
      ]
      const extension = originalExtend(modifiedStrings, ...interpolations)

      extension.displayName = StyledAsReas.displayName
      extension.styledComponentId = `${
        StyledAsReas.styledComponentId
      }--extended`

      return extension
    },
  })

  return StyledAsReas
}

export default as
