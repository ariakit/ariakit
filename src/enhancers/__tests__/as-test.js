import React from 'react'
import { shallow } from 'enzyme'
import as from '../as'

it('creates component passing property as', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} id="comp" />
  const Div = as('div')(Comp)

  const wrapper = shallow(<Div as="span" />)
  expect(wrapper.html()).toMatch(/^<span.+id="comp"/)
})

it('creates component with array of components', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} id="comp" />
  const Comp2 = ({ as: T, ...props }) => <T {...props} id="comp2" />
  const P = as([Comp2, 'p'])(Comp)

  const wrapper = shallow(<P />)
  expect(wrapper.html()).toMatch(/^<p.+id="comp2"/)
})

it('creates component passing property as with array of components', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} id="comp" />
  const Comp2 = ({ as: T, ...props }) => <T {...props} id="comp2" />
  const P = as('p')(Comp)

  const wrapper = shallow(<P as={[Comp2, 'span']} />)
  expect(wrapper.html()).toMatch(/^<span.+id="comp2"/)
})

it('creates component passing property as with another component', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} id="comp" />
  const Comp2 = ({ as: T, ...props }) => <T {...props} id="comp2" />
  const P = as('p')(Comp)
  const Div = as('div')(Comp2)

  const wrapper = shallow(<P as={Div} />)
  expect(wrapper.html()).toMatch(/^<div.+id="comp2"/)
})

it('creates component using as static method', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} id="comp" />
  const Div = as('div')(Comp)
  const P = Div.as('p')

  const wrapper = shallow(<P />)
  expect(wrapper.html()).toMatch(/^<p.+id="comp"/)
})

it('creates component using as static method with array of components', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} id="comp" />
  const Comp2 = ({ as: T, ...props }) => <T {...props} id="comp2" />
  const Div = as('div')(Comp)
  const P = Div.as([Comp2, 'p'])

  const wrapper = shallow(<P />)
  expect(wrapper.html()).toMatch(/^<p.+id="comp2"/)
})

it('renders with style', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} />
  const Div = as('div')(Comp)

  const wrapper = shallow(<Div position="absolute" />)
  expect(wrapper.html()).toMatch(/style="position:absolute/)
})

it('renders with style passing property as with another component', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} />
  const Comp2 = ({ as: T, ...props }) => <T {...props} id="comp2" />
  const P = as('p')(Comp)
  const Div = as('div')(Comp2)

  const wrapper = shallow(<P as={Div} position="absolute" />)
  expect(wrapper.html()).toMatch(/<div.+style="position:absolute/)
})

it('renders with dangerouslySetInnerHTML', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} />
  const Div = as('div')(Comp)
  const props = { dangerouslySetInnerHTML: { __html: '<b>Hello</b>' } }

  const wrapper = shallow(<Div {...props} />)
  expect(wrapper.html()).toMatch(/><b>Hello<\/b><\/div>/)
})

it('renders SVG element', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} />
  const Svg = as('svg')(Comp)

  const wrapper = shallow(<Svg as="circle" />)
  expect(wrapper.html()).toMatch(/^<circle class=/)
})

it('renders with shorthand style', () => {
  const Comp = ({ as: T, ...props }) => <T {...props} />
  const Div = as('div')(Comp)

  const wrapper = shallow(<Div width={50} />)
  expect(wrapper.html()).toMatch(/style="width:50px"/)
})
