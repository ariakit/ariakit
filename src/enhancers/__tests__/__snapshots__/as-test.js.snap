// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates component 1`] = `
Array [
  "styled(Base.as(span))",
  "Base.as(span)",
  "Reas",
  "Base",
  "Next",
  "Reas",
  "span",
]
`;

exports[`creates component 2`] = `
<styled(Base.as(span))>
  <Base.as(span)
    className="Base gzgaGM"
  >
    <Reas
      as={
        Array [
          [Function],
          "span",
        ]
      }
      className="Base gzgaGM"
    >
      <Base
        as={[Function]}
        className="Base gzgaGM"
        nextAs={
          Array [
            "span",
          ]
        }
      >
        <Next
          className="Base gzgaGM"
          id="Base"
          nextAs={
            Array [
              "span",
            ]
          }
        >
          <Reas
            as={
              Array [
                "span",
              ]
            }
            className="Base gzgaGM"
            id="Base"
          >
            <span
              className="Base gzgaGM"
              id="Base"
            />
          </Reas>
        </Next>
      </Base>
    </Reas>
  </Base.as(span)>
</styled(Base.as(span))>
`;

exports[`creates component 3`] = `

<span class="Base gzgaGM"
      id="Base"
>
</span>

`;

exports[`creates component by extending another one with extend 1`] = `
Array [
  "styled(Derivative.as(div))",
  "Derivative.as(div)",
  "Reas",
  "Derivative",
  "Base.as(span)",
  "Reas",
  "Base",
  "Next",
  "Reas",
  "Next",
  "Reas",
  "div",
]
`;

exports[`creates component by extending another one with extend 2`] = `
<styled(Derivative.as(div))>
  <Derivative.as(div)
    className="Derivative gzgaGM"
  >
    <Reas
      as={
        Array [
          [Function],
          "div",
        ]
      }
      className="Derivative gzgaGM"
    >
      <Derivative
        as={[Function]}
        className="Derivative gzgaGM"
        nextAs={
          Array [
            "div",
          ]
        }
      >
        <Base.as(span)
          as={[Function]}
          className="Derivative gzgaGM Base--extended dCNizi"
          nextAs={
            Array [
              "div",
            ]
          }
        >
          <Reas
            as={
              Array [
                [Function],
                "span",
                [Function],
                "div",
              ]
            }
            className="Derivative gzgaGM Base--extended dCNizi"
          >
            <Base
              as={[Function]}
              className="Derivative gzgaGM Base--extended dCNizi"
              nextAs={
                Array [
                  [Function],
                  "div",
                ]
              }
            >
              <Next
                className="Derivative gzgaGM Base--extended dCNizi"
                id="Base"
                nextAs={
                  Array [
                    [Function],
                    "div",
                  ]
                }
              >
                <Reas
                  as={
                    Array [
                      [Function],
                      "div",
                    ]
                  }
                  className="Derivative gzgaGM Base--extended dCNizi"
                  id="Base"
                >
                  <Next
                    as={[Function]}
                    className="Derivative gzgaGM Base--extended dCNizi"
                    id="Base"
                    nextAs={
                      Array [
                        "div",
                      ]
                    }
                  >
                    <Reas
                      as={
                        Array [
                          "div",
                        ]
                      }
                      className="Derivative gzgaGM Base--extended dCNizi"
                      id="Base"
                    >
                      <div
                        className="Derivative gzgaGM Base--extended dCNizi"
                        id="Base"
                      />
                    </Reas>
                  </Next>
                </Reas>
              </Next>
            </Base>
          </Reas>
        </Base.as(span)>
      </Derivative>
    </Reas>
  </Derivative.as(div)>
</styled(Derivative.as(div))>
`;

exports[`creates component by extending another one with extend 3`] = `

<div class="Derivative gzgaGM Base--extended dCNizi"
     id="Base"
>
</div>

`;

exports[`creates component by extending another one with styled 1`] = `
Array [
  "styled(Derivative.as(div))",
  "Derivative.as(div)",
  "Reas",
  "Derivative",
  "styled(Base.as(span))",
  "Base.as(span)",
  "Reas",
  "Base",
  "Next",
  "Reas",
  "Next",
  "Reas",
  "div",
]
`;

exports[`creates component by extending another one with styled 2`] = `
<styled(Derivative.as(div))>
  <Derivative.as(div)
    className="Derivative gzgaGM"
  >
    <Reas
      as={
        Array [
          [Function],
          "div",
        ]
      }
      className="Derivative gzgaGM"
    >
      <Derivative
        as={[Function]}
        className="Derivative gzgaGM"
        nextAs={
          Array [
            "div",
          ]
        }
      >
        <styled(Base.as(span))
          as={[Function]}
          className="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA"
          nextAs={
            Array [
              "div",
            ]
          }
        >
          <Base.as(span)
            as={[Function]}
            className="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA Base gzgaGM"
            nextAs={
              Array [
                "div",
              ]
            }
          >
            <Reas
              as={
                Array [
                  [Function],
                  "span",
                  [Function],
                  "div",
                ]
              }
              className="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA Base gzgaGM"
            >
              <Base
                as={[Function]}
                className="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA Base gzgaGM"
                nextAs={
                  Array [
                    [Function],
                    "div",
                  ]
                }
              >
                <Next
                  className="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA Base gzgaGM"
                  id="Base"
                  nextAs={
                    Array [
                      [Function],
                      "div",
                    ]
                  }
                >
                  <Reas
                    as={
                      Array [
                        [Function],
                        "div",
                      ]
                    }
                    className="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA Base gzgaGM"
                    id="Base"
                  >
                    <Next
                      as={[Function]}
                      className="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA Base gzgaGM"
                      id="Base"
                      nextAs={
                        Array [
                          "div",
                        ]
                      }
                    >
                      <Reas
                        as={
                          Array [
                            "div",
                          ]
                        }
                        className="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA Base gzgaGM"
                        id="Base"
                      >
                        <div
                          className="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA Base"
                          id="Base"
                        />
                      </Reas>
                    </Next>
                  </Reas>
                </Next>
              </Base>
            </Reas>
          </Base.as(span)>
        </styled(Base.as(span))>
      </Derivative>
    </Reas>
  </Derivative.as(div)>
</styled(Derivative.as(div))>
`;

exports[`creates component by extending another one with styled 3`] = `

<div class="Derivative gzgaGM as-test__Derivative-dFAIVE bDuygA Base"
     id="Base"
>
</div>

`;

exports[`creates component passing property as 1`] = `
Array [
  "styled(Base.as(span))",
  "Base.as(span)",
  "Reas",
  "Base",
  "Next",
  "Reas",
  "div",
]
`;

exports[`creates component passing property as 2`] = `
<styled(Base.as(span))
  as="div"
>
  <Base.as(span)
    as="div"
    className="Base gzgaGM"
  >
    <Reas
      as={
        Array [
          [Function],
          "span",
          "div",
        ]
      }
      className="Base gzgaGM"
    >
      <Base
        as={[Function]}
        className="Base gzgaGM"
        nextAs={
          Array [
            "div",
          ]
        }
      >
        <Next
          className="Base gzgaGM"
          id="Base"
          nextAs={
            Array [
              "div",
            ]
          }
        >
          <Reas
            as={
              Array [
                "div",
              ]
            }
            className="Base gzgaGM"
            id="Base"
          >
            <div
              className="Base gzgaGM"
              id="Base"
            />
          </Reas>
        </Next>
      </Base>
    </Reas>
  </Base.as(span)>
</styled(Base.as(span))>
`;

exports[`creates component passing property as 3`] = `

<div class="Base gzgaGM"
     id="Base"
>
</div>

`;

exports[`creates component passing property as with another component 1`] = `
Array [
  "styled(Base.as(span))",
  "Base.as(span)",
  "Reas",
  "Base",
  "Next",
  "Reas",
  "Derivative",
  "div",
]
`;

exports[`creates component passing property as with another component 2`] = `
<styled(Base.as(span))
  as={[Function]}
>
  <Base.as(span)
    as={[Function]}
    className="Base gzgaGM"
  >
    <Reas
      as={
        Array [
          [Function],
          "span",
          [Function],
        ]
      }
      className="Base gzgaGM"
    >
      <Base
        as={[Function]}
        className="Base gzgaGM"
        nextAs={
          Array [
            [Function],
          ]
        }
      >
        <Next
          className="Base gzgaGM"
          id="Base"
          nextAs={
            Array [
              [Function],
            ]
          }
        >
          <Reas
            as={
              Array [
                [Function],
              ]
            }
            className="Base gzgaGM"
            id="Base"
          >
            <Derivative
              className="Base gzgaGM"
              id="Base"
            >
              <div
                className="Base gzgaGM"
                id="Base"
              />
            </Derivative>
          </Reas>
        </Next>
      </Base>
    </Reas>
  </Base.as(span)>
</styled(Base.as(span))>
`;

exports[`creates component passing property as with another component 3`] = `

<div class="Base gzgaGM"
     id="Base"
>
</div>

`;

exports[`creates component passing property as with array of components 1`] = `
Array [
  "styled(Base.as(span))",
  "Base.as(span)",
  "Reas",
  "Base",
  "Next",
  "Reas",
  "Derivative",
  "Next",
  "Reas",
  "p",
]
`;

exports[`creates component passing property as with array of components 2`] = `
<styled(Base.as(span))
  as={
    Array [
      [Function],
      "p",
    ]
  }
>
  <Base.as(span)
    as={
      Array [
        [Function],
        "p",
      ]
    }
    className="Base gzgaGM"
  >
    <Reas
      as={
        Array [
          [Function],
          "span",
          [Function],
          "p",
        ]
      }
      className="Base gzgaGM"
    >
      <Base
        as={[Function]}
        className="Base gzgaGM"
        nextAs={
          Array [
            [Function],
            "p",
          ]
        }
      >
        <Next
          className="Base gzgaGM"
          id="Base"
          nextAs={
            Array [
              [Function],
              "p",
            ]
          }
        >
          <Reas
            as={
              Array [
                [Function],
                "p",
              ]
            }
            className="Base gzgaGM"
            id="Base"
          >
            <Derivative
              as={[Function]}
              className="Base gzgaGM"
              id="Base"
              nextAs={
                Array [
                  "p",
                ]
              }
            >
              <Next
                className="Base gzgaGM"
                id="Derivative"
                nextAs={
                  Array [
                    "p",
                  ]
                }
              >
                <Reas
                  as={
                    Array [
                      "p",
                    ]
                  }
                  className="Base gzgaGM"
                  id="Derivative"
                >
                  <p
                    className="Base gzgaGM"
                    id="Derivative"
                  />
                </Reas>
              </Next>
            </Derivative>
          </Reas>
        </Next>
      </Base>
    </Reas>
  </Base.as(span)>
</styled(Base.as(span))>
`;

exports[`creates component passing property as with array of components 3`] = `

<p class="Base gzgaGM"
   id="Derivative"
>
</p>

`;

exports[`creates component using as static method 1`] = `
Array [
  "styled(styled(Base.as(span)).as(p))",
  "styled(Base.as(span)).as(p)",
  "Reas",
  "styled(Base.as(span))",
  "Base.as(span)",
  "Reas",
  "Base",
  "Next",
  "Reas",
  "Next",
  "Reas",
  "p",
]
`;

exports[`creates component using as static method 2`] = `
<styled(styled(Base.as(span)).as(p))>
  <styled(Base.as(span)).as(p)
    className="Base gzgaGM"
  >
    <Reas
      as={
        Array [
          [Function],
          "p",
        ]
      }
      className="Base gzgaGM"
    >
      <styled(Base.as(span))
        as={[Function]}
        className="Base gzgaGM"
        nextAs={
          Array [
            "p",
          ]
        }
      >
        <Base.as(span)
          as={[Function]}
          className="Base gzgaGM Base gzgaGM"
          nextAs={
            Array [
              "p",
            ]
          }
        >
          <Reas
            as={
              Array [
                [Function],
                "span",
                [Function],
                "p",
              ]
            }
            className="Base gzgaGM Base gzgaGM"
          >
            <Base
              as={[Function]}
              className="Base gzgaGM Base gzgaGM"
              nextAs={
                Array [
                  [Function],
                  "p",
                ]
              }
            >
              <Next
                className="Base gzgaGM Base gzgaGM"
                id="Base"
                nextAs={
                  Array [
                    [Function],
                    "p",
                  ]
                }
              >
                <Reas
                  as={
                    Array [
                      [Function],
                      "p",
                    ]
                  }
                  className="Base gzgaGM Base gzgaGM"
                  id="Base"
                >
                  <Next
                    as={[Function]}
                    className="Base gzgaGM Base gzgaGM"
                    id="Base"
                    nextAs={
                      Array [
                        "p",
                      ]
                    }
                  >
                    <Reas
                      as={
                        Array [
                          "p",
                        ]
                      }
                      className="Base gzgaGM Base gzgaGM"
                      id="Base"
                    >
                      <p
                        className="Base gzgaGM"
                        id="Base"
                      />
                    </Reas>
                  </Next>
                </Reas>
              </Next>
            </Base>
          </Reas>
        </Base.as(span)>
      </styled(Base.as(span))>
    </Reas>
  </styled(Base.as(span)).as(p)>
</styled(styled(Base.as(span)).as(p))>
`;

exports[`creates component using as static method 3`] = `

<p class="Base gzgaGM"
   id="Base"
>
</p>

`;

exports[`creates component using as static method with array of components 1`] = `
Array [
  "styled(styled(Base.as(span)).as(Derivative,p))",
  "styled(Base.as(span)).as(Derivative,p)",
  "Reas",
  "styled(Base.as(span))",
  "Base.as(span)",
  "Reas",
  "Base",
  "Next",
  "Reas",
  "Next",
  "Reas",
  "Derivative",
  "Next",
  "Reas",
  "p",
]
`;

exports[`creates component using as static method with array of components 2`] = `
<styled(styled(Base.as(span)).as(Derivative,p))>
  <styled(Base.as(span)).as(Derivative,p)
    className="Base-Derivative gzgaGM"
  >
    <Reas
      as={
        Array [
          [Function],
          [Function],
          "p",
        ]
      }
      className="Base-Derivative gzgaGM"
    >
      <styled(Base.as(span))
        as={[Function]}
        className="Base-Derivative gzgaGM"
        nextAs={
          Array [
            [Function],
            "p",
          ]
        }
      >
        <Base.as(span)
          as={[Function]}
          className="Base-Derivative gzgaGM Base gzgaGM"
          nextAs={
            Array [
              [Function],
              "p",
            ]
          }
        >
          <Reas
            as={
              Array [
                [Function],
                "span",
                [Function],
                [Function],
                "p",
              ]
            }
            className="Base-Derivative gzgaGM Base gzgaGM"
          >
            <Base
              as={[Function]}
              className="Base-Derivative gzgaGM Base gzgaGM"
              nextAs={
                Array [
                  [Function],
                  [Function],
                  "p",
                ]
              }
            >
              <Next
                className="Base-Derivative gzgaGM Base gzgaGM"
                id="Base"
                nextAs={
                  Array [
                    [Function],
                    [Function],
                    "p",
                  ]
                }
              >
                <Reas
                  as={
                    Array [
                      [Function],
                      [Function],
                      "p",
                    ]
                  }
                  className="Base-Derivative gzgaGM Base gzgaGM"
                  id="Base"
                >
                  <Next
                    as={[Function]}
                    className="Base-Derivative gzgaGM Base gzgaGM"
                    id="Base"
                    nextAs={
                      Array [
                        [Function],
                        "p",
                      ]
                    }
                  >
                    <Reas
                      as={
                        Array [
                          [Function],
                          "p",
                        ]
                      }
                      className="Base-Derivative gzgaGM Base gzgaGM"
                      id="Base"
                    >
                      <Derivative
                        as={[Function]}
                        className="Base-Derivative gzgaGM Base gzgaGM"
                        id="Base"
                        nextAs={
                          Array [
                            "p",
                          ]
                        }
                      >
                        <Next
                          className="Base-Derivative gzgaGM Base gzgaGM"
                          id="Derivative"
                          nextAs={
                            Array [
                              "p",
                            ]
                          }
                        >
                          <Reas
                            as={
                              Array [
                                "p",
                              ]
                            }
                            className="Base-Derivative gzgaGM Base gzgaGM"
                            id="Derivative"
                          >
                            <p
                              className="Base-Derivative gzgaGM Base"
                              id="Derivative"
                            />
                          </Reas>
                        </Next>
                      </Derivative>
                    </Reas>
                  </Next>
                </Reas>
              </Next>
            </Base>
          </Reas>
        </Base.as(span)>
      </styled(Base.as(span))>
    </Reas>
  </styled(Base.as(span)).as(Derivative,p)>
</styled(styled(Base.as(span)).as(Derivative,p))>
`;

exports[`creates component using as static method with array of components 3`] = `

<p class="Base-Derivative gzgaGM Base"
   id="Derivative"
>
</p>

`;

exports[`creates component with array of components 1`] = `
Array [
  "styled(Base.as(Derivative,p))",
  "Base.as(Derivative,p)",
  "Reas",
  "Base",
  "Next",
  "Reas",
  "Derivative",
  "Next",
  "Reas",
  "p",
]
`;

exports[`creates component with array of components 2`] = `
<styled(Base.as(Derivative,p))>
  <Base.as(Derivative,p)
    className="Base-Derivative gzgaGM"
  >
    <Reas
      as={
        Array [
          [Function],
          [Function],
          "p",
        ]
      }
      className="Base-Derivative gzgaGM"
    >
      <Base
        as={[Function]}
        className="Base-Derivative gzgaGM"
        nextAs={
          Array [
            [Function],
            "p",
          ]
        }
      >
        <Next
          className="Base-Derivative gzgaGM"
          id="Base"
          nextAs={
            Array [
              [Function],
              "p",
            ]
          }
        >
          <Reas
            as={
              Array [
                [Function],
                "p",
              ]
            }
            className="Base-Derivative gzgaGM"
            id="Base"
          >
            <Derivative
              as={[Function]}
              className="Base-Derivative gzgaGM"
              id="Base"
              nextAs={
                Array [
                  "p",
                ]
              }
            >
              <Next
                className="Base-Derivative gzgaGM"
                id="Derivative"
                nextAs={
                  Array [
                    "p",
                  ]
                }
              >
                <Reas
                  as={
                    Array [
                      "p",
                    ]
                  }
                  className="Base-Derivative gzgaGM"
                  id="Derivative"
                >
                  <p
                    className="Base-Derivative gzgaGM"
                    id="Derivative"
                  />
                </Reas>
              </Next>
            </Derivative>
          </Reas>
        </Next>
      </Base>
    </Reas>
  </Base.as(Derivative,p)>
</styled(Base.as(Derivative,p))>
`;

exports[`creates component with array of components 3`] = `

<p class="Base-Derivative gzgaGM"
   id="Derivative"
>
</p>

`;

exports[`renders SVG element 1`] = `

<circle id="Base"
        class="Base gzgaGM"
>
</circle>

`;

exports[`renders with dangerouslySetInnerHTML 1`] = `

<span class="Base gzgaGM"
      id="Base"
>
  <b>
    Hello
  </b>
</span>

`;

exports[`renders with style 1`] = `

<span class="Base gzgaGM"
      id="Base"
      style="position:absolute"
>
</span>

`;

exports[`renders with style passing property as with another component 1`] = `

<div class="Base gzgaGM Derivate"
     id="Derivate"
     style="position:absolute"
>
</div>

`;
